### PR TITLE
feat(archive/unarchive): add archive/unarchive mutation by URL

### PR DESCRIPTION
**Pull request by kschelonka**
## Goal
Add archive/unarchive mutations for SavedItems by Url, to support offline mobile clients

## I'd love feedback/perspectives on:
- any

## Implementation Decisions
- Combining archive/unarchive together since they're so similar
- Moving shared code to a new SavedItem model
- Continuing pattern of using Parser to resolve the SavedItem url, until the givenUrl field can be indexed in `list` table, to avoid potentially huge lookups when users have millions of items

## References

Jira ticket:
* [IN-1472]
* [IN-1476]

[IN-1472]: https://getpocket.atlassian.net/browse/IN-1472?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[IN-1476]: https://getpocket.atlassian.net/browse/IN-1476?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/dataService/savedItemsService.ts [hunk 1]
```diff
@@ -117,7 +117,7 @@ export class SavedItemDataService {
    * Fetch a single SavedItem from a User's list
    * @param itemId the savedItem ID to fetch
    */
-  public getSavedItemById(itemId: string): Promise<SavedItem> {
+  public getSavedItemById(itemId: string): Promise<SavedItem | null> {
```
Comment (kschelonka) [inline src/dataService/savedItemsService.ts:120]: Just being more explicit

File: src/resolvers/mutation.ts [hunk 2]
```diff
@@ -134,13 +134,7 @@ export async function updateSavedItemArchive(
...
   context: IContext
 ): Promise<SavedItem> {
-  const savedItemService = new SavedItemDataService(context);
-  const savedItem = await savedItemService.updateSavedItemArchiveProperty(
-    args.id,
-    true
-  );
-  context.emitItemEvent(EventType.ARCHIVE_ITEM, savedItem);
-  return savedItem;
+  return context.models.savedItem.archiveById(args.id);
```
Comment (kschelonka) [inline src/resolvers/mutation.ts:137]: These were essentially copy-pasted moved to the SavedItem model

File: src/models/SavedItem.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+   * https://getpocket.atlassian.net/browse/IN-1478
+   * @returns the itemId associated with the url
+   * @throws NotFound Error if the itemId does not exist
+   *  for the URL in the parser service; do not trigger a parse
+   *  to avoid any risk of IDs getting out of sync/multiple IDs
+   *  for a savedItem record, etc. The ID should already exist
+   *  if update mutations are being called on the SavedItem entity.
+   */
+  private async fetchIdFromUrl(url: string): Promise<string> {
+    const id = await ParserCaller.getItemIdFromUrl(url);
```
Comment (sri-kirsh) [inline src/models/SavedItem.ts:107]: 👍

File: src/test/graphql/mutations/savedItem/savedItemTag.integration.ts [hunk 4]
```diff
@@ -1,14 +1,14 @@
 import { ApolloServer } from '@apollo/server';
-import { ContextManager } from '../../../server/context';
-import { readClient } from '../../../database/client';
-import { startServer } from '../../../server/apollo';
-import { mockParserGetItemIdRequest } from '../../utils/parserMocks';
+import { ContextManager } from '../../../../server/context';
```
Comment (kschelonka) [inline src/test/graphql/mutations/savedItem/savedItemTag.integration.ts:2]: Just doing some folder organization

File: src/test/graphql/mutations/savedItem/savedItemUnArchive.integration.ts [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    expect(eventSpy.firstCall.args).toStrictEqual([
+      'UNARCHIVE_ITEM',
+      expect.objectContaining({
+        url: givenUrl,
+        time_updated: new Date(testTimestamp),
+      }),
+    ]);
+  });
+  // This might change to a no-op later, but let's stick with current behavior of
+  // updatesavedItemUnArchive and have it documented in tests
```
Comment (sri-kirsh) [inline src/test/graphql/mutations/savedItem/savedItemUnArchive.integration.ts:169]: make sense

**Review by sri-kirsh — APPROVED** [review]

**Pull request merged**